### PR TITLE
analyzer: move InlineTableScan into Analyzer.

### LINKED ISSUE #5682: Move InlineTableScan and code into Analyzer
### Is your feature request related to a problem or challenge?

part of #5681 

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #5682

# Rationale for this change

# What changes are included in this PR?

 move InlineTableScan into Analyzer.

# Are these changes tested?

Add Unit test for Analyzer

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jackwener** [comment]
cc @alamb @mingmwang @liukun4515

**Review by alamb — COMMENTED** [review]
Taking a step back, I thought the purpose of the Analyzer was to do semantic analysis (things like doing type checks) or maybe coercing the types so the the plan became valid.

It isn't clear to me that `InlineTableScan` falls into this category

However,  the other rule,  https://github.com/apache/arrow-datafusion/blob/main/datafusion/optimizer/src/analyzer/count_wildcard_rule.rs does not fot this criteria either 🤔 

So therefore I probably misunderstand what is supposed to go in the `Analyzer` and what should remain in `Optimizer`. Is the distinction between the two documented somewhere?

**Comment by jackwener** [comment]
> However, the other rule, https://github.com/apache/arrow-datafusion/blob/main/datafusion/optimizer/src/analyzer/count_wildcard_rule.rs does not fot this criteria either 🤔

Spark has a similar rule `ExtractWindowExpressions` like it.

> Taking a step back, I thought the purpose of the Analyzer was to do semantic analysis (things like doing type checks) or maybe coercing the types so the the plan became valid.

I think any action to make plan valid should be put into analyzer like `resolvePlan`, `typeCoercion`, .....
I think `inlineTableScan` also belong this category.

**Comment by alamb** [comment]
> I think any action to make plan valid should be put into analyzer like resolvePlan, typeCoercion, .....
> I think inlineTableScan/count_wildcard_rule also belong this category.

That makes sense to me. I tried to capture this in doc comments for future readers here:  https://github.com/apache/arrow-datafusion/pull/5705

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs [hunk 1]
```diff
@@ -15,73 +15,113 @@
...
-                    Ok(Some(plan.build()?))
-                } else {
-                    Ok(None)
-                }
-            }
-            _ => Ok(None),
-        }
+impl AnalyzerRule for InlineTableScan {
+    fn analyze(&self, plan: &LogicalPlan, _: &ConfigOptions) -> Result<LogicalPlan> {
+        plan.clone().transform_up(&analyze_internal)
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/inline_table_scan.rs:45]: The clone is unfortunate, but a legacy of the existing API

**Review by alamb — APPROVED** [review]
Thanks @jackwener  -- makes sense to me

**Pull request merged**